File: SBiswas/Problems/Milestone1/Problem3.cpp
#include <iostream>

/*
    TODO: Modularize the code. For example build a different header file 
    with class Node and Class BST for binary search tree.
*/

class Node {
    
    public:
    int data;
    Node* left;
    Node* right;
    
    Node(int val)
    {
        data = val;
        left = nullptr;
        right = nullptr;
    }
    
};

Node* insert_in_BST(Node* root, int val)
{   
    /*
    Insert Function for Binary Search Tree
    */

   if(root == nullptr)
   {
        Node* temp = new Node(val);
        return temp;
   }

    if(val > root->data)
    {
        // Go to right of the root
        root->right = insert_in_BST(root->right,val);
    }
    else if (val < root->data)
    {
        // Go to left of the root
        root->left = insert_in_BST(root->left,val);
    }    
    
    return root;
}

void preorder(Node* root)
{
    if(root == nullptr)
    {
        return;
    }
    
    std::cout<<root->data<<" - ";
    preorder(root->left);
    preorder(root->right);
}

void postorder(Node* root)
{
    if(root == nullptr)
    {
        return;
    }
    
    postorder(root->left);
    postorder(root->right);
    std::cout<<root->data<<" - ";
}

void inorder(Node* root)
{
    if(root == nullptr)
    {
        return;
    }
    
    inorder(root->left);
    std::cout<<root->data<<" - ";
    inorder(root->right);
}


int main()
{
    Node* my_tree = nullptr; 
    
    // insert(my_tree,1);
    my_tree = insert_in_BST(my_tree,5);
    my_tree = insert_in_BST(my_tree,6);
    my_tree = insert_in_BST(my_tree,7);
    
    my_tree = insert_in_BST(my_tree,1);
    my_tree = insert_in_BST(my_tree,9);

    my_tree = insert_in_BST(my_tree,4);
    my_tree = insert_in_BST(my_tree,0);
    my_tree = insert_in_BST(my_tree,8);
    
    std::cout<<"Inorder Traversal: \n";
    inorder(my_tree);
    std::cout<<'\n';
    std::cout<<"Preorder Traversal: \n";
    preorder(my_tree);
    std::cout<<'\n';
    std::cout<<"Postorder Traversal: \n";
    postorder(my_tree);
    std::cout<<'\n';
    std::cout<<"Inorder Traversal outputs the nodes in ascending order.\n";
    
    return 0;
}

File: VishnuThakral/Milestone1/Problem3.py
# Create a Binary Search Tree and Do Pre/In/Post Oder traversal of the tree
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class BST:
    def __init__(self):
        self.root = None
        self.arr = []

    def insert_node(self,node,val):
        if node.val > val:
            if node.left is None:
                node.left = TreeNode(val)
            else:
                self.insert_node(node.left,val)
        else:
            if node.right is None:
                node.right = TreeNode(val)
            else:
                self.insert_node(node.right,val)
    def add_node(self, val):
        if self.root is None:
            self.root = TreeNode(val)
        else:
            self.insert_node(self.root,val)

    def _in_order(self,node):
        if node:
            if node.left:
                self._in_order(node.left)
            self.arr.append(node.val)
            if node.right:
                self._in_order(node.right)

    def in_order(self):
        self.arr = []
        self._in_order(self.root)
        return self.arr

    def _pre_order(self,node):
        if node:
            self.arr.append(node.val)
            if node.left:
                self._pre_order(node.left)
            if node.right:
                self._pre_order(node.right)

    def pre_order(self):
        self.arr = []
        self._pre_order(self.root)
        return self.arr

    def _post_order(self,node):
        if node:
            if node.left:
                self._post_order(node.left)
            if node.right:
                self._post_order(node.right)
            self.arr.append(node.val)

    def post_order(self):
        self.arr = []
        self._post_order(self.root)
        return self.arr

arr = [3,1,2,5,6,4]
bst = BST()
for i in arr:
    bst.add_node(i)
print(bst.in_order())
print(bst.pre_order())
print(bst.post_order())
